Remove items with no url prop when saving to S3 on publish

lennym opened the pull request:
We want to be able to send items back into the results pipe to the client without also necessarily saving them to s3 as well. In particular, it does not make sense to save items to s3 which do not url.

As such, we remove items from the items array which have no url property before saving to s3.

Also slightly refactor the tests for this method to more explicitly define the calls to downstream AWS services by stubbing the AWS SDK.

lennym committed:
Remove items with n url prop when saving to S3 on publish

We want to be able to send items back into the results pipe to the client without also necessarily saving them to s3 as well. In particular, it does not make sense to save items to s3 which do not url.

As such, we remove items from the items array which have no url property before saving to s3.

Also slightly refactor the tests for this method to more explicitly define the calls to downstream AWS services by stubbing the AWS SDK.

lennym committed:
2.17.0

lennym commented:
On further thought, we might want to be more explicit and filter items based on the `type` property. Since we're really saving packages and articles for reference later, so a `type: 'package'` or `type: 'tile'` filter might be more appropriate that a slightly arbitrary url filter..

nelsonic commented:
@lennym looks good. merging. 👍

nelsonic commented:
@lennym that would require us to add a `type` property to each item that we want to store on S3. 
I don't have an _objection_ to doing that but it is more work (_in 3 places_), whereas your `url` filter does the job already.

lennym commented:
@nelsonic I think those items have those type properties already as far as I can tell. Happy to merge as is though if you are.

lennym commented:
I stand corrected. Tiles and filters have a type property, packages do not.

nelsonic commented:
@lennym you're *right*, [lambda-tile-provider/saveHandler.js#L38](https://github.com/numo-labs/lambda-tile-provider/blob/73d3de454d4979cfcaa45f57f11cf3742dcd1885/lib/saveHandler.js#L38) *does* specify its `type` and we can *easily* add it to the packages...

nelsonic commented:
@lennym please take a look at: https://github.com/numo-labs/lambda-ne-classic-package-provider/pull/76 does this allow us to perform the more *explicit* check now in `saveRecord` ?

lennym committed:
Refactor to filter based on item type

Filtering on URL is fairly arbitrary, where what we actually wish to do is store tiles and packages for later retreival base don url. As such, change the filter from being based on url property to a test for type.

lennym commented:
@nelsonic Updated to filter based on `type` rather than `url`.

nelsonic commented:
@jruts what `type` of record are you planning on assigning for the records stored by `lambda-search-request-handler` for the `neo4j` response?

nelsonic commented on lib/index.js:
```diff
@@ -312,9 +312,14 @@ var resultBucket = new AWS.S3({ params:
-  // console.log(params.items);
-  params.items.forEach(function (item) {
+  const filtered = params.items.filter((item) => {
+    return item.type && item.type.match(/^tile|package$/);
```
@jruts we will need to add the additional types here for your `meta` files.

nelsonic commented:
looks good. let's merge! 👍
